awscrt/aio/aiohttp.py

```diff
@@ -0,0 +1,434 @@
+
+    async def wait_for_completion(self) -> int:
+        """Wait asynchronously for the stream to complete.
+
+        Returns:
+            int: The response status code.
+        """
+        return await self._completion_future
+
+    async def _set_request_body_generator(self, body_iterator: AsyncIterator[bytes]):
+        ...
+
+
+class AIOHttpClientStream(AIOHttpClientStreamUnified):
+    """Async HTTP stream that sends a request and receives a response.
+
+    Create an AIOHttpClientStream with `AIOHttpClientConnection.request()`.
+
+    Attributes:
+        connection (AIOHttpClientConnection): This stream's connection.
+
+        completion_future (asyncio.Future): Future that will contain
+            the response status code (int) when the request/response exchange
+            completes. If the exchange fails to complete, the Future will
+            contain an exception indicating why it failed.
+
+    Notes:
+        All async method on a stream (await stream.next(), etc.) must be performed in the
+        thread that owns the event loop used to create the stream
+    """
+
+    def __init__(self, connection: AIOHttpClientConnection, request: HttpRequest,
+                 loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
+        """Initialize an HTTP client stream.
+
+        Args:
+            connection (AIOHttpClientConnection): The connection to send the request on.
+            request (HttpRequest): The HTTP request to send.
+            loop (Optional[asyncio.AbstractEventLoop]): Event loop to use for async operations.
+                If None, the current event loop is used.
+        """
+        super().__init__(connection, request, loop=loop)
+
+
+class AIOHttp2ClientStream(AIOHttpClientStreamUnified):
+    """HTTP/2 stream that sends a request and receives a response.
+
+    Create an AIOHttp2ClientStream with `AIOHttp2ClientConnection.request()`.
+    """
+
+    def __init__(self,
+                 connection: AIOHttpClientConnection,
+                 request: HttpRequest,
+                 request_body_generator: AsyncIterator[bytes] = None,
+                 loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
+        super().__init__(connection, request, request_body_generator=request_body_generator, loop=loop)
+
+    async def _write_data(self, body, end_stream):
+        future = Future()
+        body_stream = InputStream.wrap(body, allow_none=True)
+
+        def on_write_complete(error_code: int) -> None:
+            if future.cancelled():
+                # the future was cancelled, so we don't need to set the result or exception
+                return
+            if error_code:
+                future.set_exception(awscrt.exceptions.from_code(error_code))
+            else:
+                future.set_result(None)
+
+        _awscrt.http2_client_stream_write_data(self, body_stream, end_stream, on_write_complete)
+        await asyncio.wrap_future(future)
```

Comment:
why do we want to use concurrent Future here and then wrap it into asyncio one? why not use async io one from start?

Reply:
Since the concurrent future is thread safe, but the asyncio one is not.

If we just get asyncio future, we will need to invoke `self._loop.call_soon_threadsafe()` to set the future. Either way, it's a bit awkward.

But, yeah, I did use the asyncio future directly for some other places, I guess it would be nice to keep it consistent.